stdlib/public/core/Rotate.swift

```diff
@@ -0,0 +1,716 @@
+    @discardableResult
+    public mutating func rotate(toFirst middle: Index) -> Index {
+        switch middle {
+        case ...startIndex, endIndex...:
+            // A base case, otherwise the true/false and false/true cases of the
+            // switch below would trigger infinite recursion.
+            return startIndex
+        default:
+            // With the given index as the pivot between partitions, switch the
+            // order of said partitions.  Do it without testing the relative
+            // lengths in advance.  If the partitions have different lengths,
+            // the prefix/whole part of the originally-second partition is
+            // finalized and the remainder (the suffix of the longer original
+            // partition and the whole/prefix part of the originally-first
+            // partition) needs to be rotated.
+            var i = startIndex, j = middle
+            while true {
+                switch (i == middle, j == endIndex) {
+                case (false, false):
+                    // Move one step closer by doing the core swap.
+                    swapAt(i, j)
+                    formIndex(after: &i)
+                    formIndex(after: &j)
+                case (false, true):
+                    // The first partition is longer than the second.
+                    self[i...].rotate(toFirst: middle)
+                    return i
+                case (true, false):
+                    // The first partition is shorter than the second.
+                    return self[middle...].rotate(toFirst: j)
+                case (true, true):
+                    // The partitions have the same length.
+                    return middle
+                }
+            }
+        }
+    }
+
+    /// Rotates the elements of the collection such that the value at the given
+    /// index moves rightward to just before `endIndex`.
+    ///
+    /// Passing `endIndex` or its immediate predecessor (if any) as `middle` has
+    /// no effect.
+    ///
+    /// The following example rotates the elements of an array of characters so
+    /// the second character moves to be the last:
+    ///
+    ///     var café: [Character] = ["C", "a", "f", "é"]
+    ///     café.rotate(toLast: café.index(after: café.startIndex))
+    ///     print(café)
+    ///     // Prints "["f", "é", "C", "a"]"
+    ///
+    /// - Precondition: `middle` must be a valid index of this collection.
+    ///
+    /// - Parameter middle: The index of the element whose value will move to
+    ///   just before `endIndex`, unless it's `endIndex` (then it's ignored).
+    ///
+    /// - Returns: The new index of the value originally at `startIndex`.
+    ///
+    /// - Postcondition: The collection is a right-rotation of its old state,
+    ///   *i.e.* `newValue` equals `oldValue[middle>..] + oldValue[...middle]`.
+    ///
+    /// - Complexity: O(*n*), where *n* is the number of elements in the
+    ///   collection.
+    @inlinable
```

Comment:
Why is this `@inlinable` and `rotate(toFirst:)` is not?

Reply:
Because `rotate(toLast:)` is a trivial redirect to `rotate(toFirst:)`, but that doesn't mean that the latter's implementation is trivial.